src/flash.rs

```diff
@@ -224,22 +255,48 @@ impl<'a, const SECTOR_SZ_KB: u32> FlashWriter<'a, SECTOR_SZ_KB> {
         // Unlock Flash
         self.unlock()?;
 
-        for idx in (0..data.len()).step_by(2) {
+        // According to RM0440 Rev 7, "It is only possible to program double word (2 x 32-bit data)"
+        for idx in (0..data.len()).step_by(8) {
+            // Check the starting address
             self.valid_address(offset + idx as u32)?;
+            // Check the ending address to make sure there is no overflow
+            self.valid_address(offset + 8 + idx as u32)?;
 
-            let write_address = (FLASH_START + offset + idx as u32) as *mut u16;
+            let write_address1 = (FLASH_START + offset + idx as u32) as *mut u32;
+            let write_address2 = (FLASH_START + offset + 4 + idx as u32) as *mut u32;
+
+            let word1: u32;
+            let word2: u32;
+
+            // Check if there is enough data to make 2 words, if there isn't, pad the data with 0xFF
+            if idx + 8 > data.len() {
+                let mut tmp_buffer = [255 as u8; 8];
+                for jdx in idx..data.len() {
+                    tmp_buffer[jdx] = data[idx + jdx];
+                }
+                let tmp_dword = u64::from_le_bytes(tmp_buffer);
+                word1 = tmp_dword as u32;
+                word2 = (tmp_dword >> 32) as u32;
```

Comment:
A note from my colleague:

> For me this is a bit to much intelligence/"magic" within the driver. In my opinion we should check whether the user passes data to us whose length is a multiple of 8 and return an error, if this is not the case.
> Background:
> Imagine, a user doesn't know, that we only can write multiples of 8. This user passes blocks of 4 bytes (this is valid for a lot of MCUs). The first write succeeds, because you added the missing 4 bytes. The next write will fail, because writing 0xFF is internally another state than a deleted flash block and the MCU's flash routine will return an error, if the user tries to write his next block of 4 bytes.
> We should inform the user that he does not fulfill the boundary conditions as soon as he writes the first time (via an error). This is because the error in the second write is de facto only a subsequent error.

Reply:
Yes, I see y'alls perspective. On a previous MCU I worked with, as long as the alignment was fine, a flash bits could go from 1 -> 0, so padding with 0xFF didn't cause an issue and made writing data easier since the user didn't need to worry about data length, only alignment of the start address and that the flash block was already empty. 

Let me write some test cases to cover this and see what happens. If there are any issues with this test I will add a check to ensure incoming arrays are divisible by 8 and return an error (or add a new one) that makes this condition clear. 

Thank you for the feedback.

Reply:
I did a bit more testing and recognize your colleagues complaint. However, I still would like writing data to be easy in most cases. A hopeful solution is adding a flag to `FlashWriter.write` called `force_data_padding`, along with function comments, that makes it clear that data padding is going on.